batching messages

LokiMidgard opened the pull request:
attemp to get around too complex types.

macabeus commented on src/build/header.ts:
```diff
@@ -4,18 +4,38 @@ const bannerMessage = (
 )
 
-const vanillaSupport = dedent`
+const vanillaSupport = (messagesVariables: MessageVariablesMap) => {
```
Your code works well, but would be good to follow the code style of the code base. You could check more on [`.eslintrc.js`]( https://github.com/macabeus/fluent-typescript/blob/master/.eslintrc.js) (and if you are using vscode, I really recommend to install the extension [`ESLint`](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint).

macabeus commented on src/build/type-pattern-arguments.ts:
```diff
@@ -7,27 +7,56 @@ const wrapVariables = (variables: Variables) => variables.map(i => `'${i}': Flue
+
+  const batchSize = 25;
+  let returnValue: string = '';
+  for (let index = 0; index < options.length; index++) {
```
This for has ~25 lines of code, and there are 6 operators that updates `returnValue`. It isn't good, because makes harder to read (since the values could updates by many ways).

Would be good to split these lines on smaller and pure functions, like `hasVariables` and `buildTypePatternArguments`.

LokiMidgard replied:
refactored to use map

macabeus commented on src/build/type-pattern-arguments.ts:
```diff
@@ -7,27 +7,56 @@ const wrapVariables = (variables: Variables) => variables.map(i => `'${i}': Flue
+
+
+  }
+  returnValue += dedent`
+        : never
+        )
+        
+        `
+  return returnValue;
+
+
+
+
```
Lint
```suggestion
  returnValue += dedent`
      : never
    )
  `

  return returnValue
```

macabeus commented on example-vanilla/assets/locales/pt-br/translations.ftl:
```diff
@@ -1,3 +1,56 @@
 hello = Olá { $firstName }
```
I think that isn't necessary to update the vanilla example with that. We should instead add a new test case to check that.

Also, is necessary to update the current test cases with these changes.

LokiMidgard replied:
I reverted the example

LokiMidgard commented:
I'll look into your review, I'm not sure how long it takes through.

LokiMidgard commented on src/build/header.ts:
```diff
@@ -53,22 +73,36 @@ const fluentReactSupport = dedent`
+  ${returnValue}
 
-const buildHeader = (target: TargetsSupported) => `${bannerMessage}${header(target)}\n`
+  ${targetSupport[target](messagesVariables)}
```
I try to add a new data structure, that contains the batches of message variables.

An array of an Array. That way the splitting logic is only in one place.

LokiMidgard commented on src/build/header.ts:
```diff
@@ -4,18 +4,38 @@ const bannerMessage = (
+
+  let returnValue = ''
+
+  for (let index = 0; index < numberOfBatches; index++) {
```
I refactored most loops to maps and let to const.

LokiMidgard commented on src/build/header.ts:
```diff
@@ -4,18 +4,38 @@ const bannerMessage = (
+`}
 
-const i18NextSupport = dedent`
+const i18NextSupport = (messagesVariables: MessageVariablesMap) => dedent`
```
I changed `t : <T extends MessagesKey0>(...args: PatternArguments0<T>)  => string` to `t<T extends MessagesKey0>(...args: PatternArguments0<T>)  : string` so they can use parameter overload. I think bothe statements are equivalent. correct me if I'm wrong.

LokiMidgard commented on src/build/header.ts:
```diff
@@ -28,7 +48,7 @@ const i18NextSupport = dedent`
 `
 
-const fluentReactSupport = dedent`
+const fluentReactSupport = (messagesVariables: MessageVariablesMap) => dedent`
```
I'm not sure how to change those. Would I get mutliple `LocalizedPropsPatched` types?

LokiMidgard replied:
Do you have some hints how to change the React part?

LokiMidgard commented on src/build/header.ts:
```diff
@@ -53,22 +73,36 @@ const fluentReactSupport = dedent`
-  ${targetSupport[target]}
-`
+  return dedent`
+  ${returnValue}
```
All `returnValue`s should be refactored away.

LokiMidgard commented on src/build/type-pattern-arguments.ts:
```diff
@@ -7,27 +7,56 @@ const wrapVariables = (variables: Variables) => variables.map(i => `'${i}': Flue
-            ? [T, { ${wrapVariables(variables).join(',')} }]
+  const options = Object.entries(messagesVariables);
+
+  const batchSize = 25;
```
I pulled it up in the build method so it will only be used once.

LokiMidgard commented:
I think I fixed everything besides the `fluent-react` code. I'm not sure how the classes should be structured.

I also added a new Test which is failing. But I think the test is correct and it should be no problem with the changed code.
I have messages `t1` to `t27` defined in the test, however the test also have the messages `hello`, `how-are-you` and `bye`.

It seams that the test caches somehow the messages from previous test. So either I used the test methods wrong, or there is a caching bug.

macabeus commented:
@LokiMidgard I'm sorry for delay to reply your PR. I'm working on it now, and I noticed that VSCode won't work well when `.d.ts` is so big. I opened an issue on VSCode repository: https://github.com/microsoft/vscode/issues/105662

macabeus commented:
@LokiMidgard Sorry for the long delay to review your PR, update it and merge it. I know that is boring to wait to receive a reply on a PR.
I updated your code. I recommend that you take a look on my commit, but some key points:
- fix the tests cases, and rearrange it (for example, add the optional parameter `chuckSize`)
- rename `batch` to `chuck`, to match better on some JS functions on popular libraries, such as on lodash [chuck](https://lodash.com/docs/4.17.15#chunk)
- add support on `fluent-react`
- update the examples

macabeus commented:
About the bug on vscode, it's because of my extension [vscode-fluent](https://github.com/macabeus/vscode-fluent) (I commented it here: https://github.com/microsoft/vscode/issues/105662#issuecomment-683476034).

I really don't know why, but `vscode-fluent` is trying to check the syntax on `.ts` files - but it should only work on `.ftl` files.
